### PR TITLE
New table and service to store auth tokens

### PR DESCRIPTION
**Pull request by Shazwazza**
Implements the auth token retrieval and storage for the identity implementation. This is now automatically done for providers and the back office user manager can be used to retreive and store all tokens.

Fixes locking on the config writer.

Removes the abstract NodeObjectTypeId from the base repo since it shouldn't be there.

## Testing

I have definitely tested that this passes tokens and stores them, it's a bit of a pain to set that up and test but it works :) 

Since this has some migrations, it will launch the upgrader so you'll want to make sure that all works.

Tests are passing 👍

### CODE REVIEW

File: src/Umbraco.Infrastructure/Security/UmbracoIdentityUser.cs [hunk 1]
```diff
@@ -180,6 +182,37 @@ public ICollection<IIdentityUserLogin> Logins
...
 
+        /// <summary>
+        /// Gets the external login tokens collection
+        /// </summary>
+        public ICollection<IIdentityUserToken> LoginTokens
+        {
+            get
+            {
+                // return if it exists
+                if (_tokens != null)
```
Comment (bergmania) [inline src/Umbraco.Infrastructure/Security/UmbracoIdentityUser.cs:193]:
```suggestion
                if (_tokens is not null)
```

File: src/Umbraco.Infrastructure/Security/UmbracoIdentityUser.cs [hunk 2]
```diff
@@ -180,6 +182,37 @@ public ICollection<IIdentityUserLogin> Logins
...
+                {
+                    return _tokens;
+                }
+
+                _tokens = new ObservableCollection<IIdentityUserToken>();
+
+                // if the callback is there and hasn't been created yet then execute it and populate the logins
+                if (_getTokens != null && !_getTokens.IsValueCreated)
```
Comment (bergmania) [inline src/Umbraco.Infrastructure/Security/UmbracoIdentityUser.cs:201]:
```suggestion
                if (_getTokens?.IsValueCreated != true)
```

File: src/Umbraco.Tests.Integration/Umbraco.Infrastructure/Services/ExternalLoginServiceTests.cs [hunk 3]
```diff
@@ -202,6 +203,31 @@ public void Add_Logins()
             }
         }
 
+        [Test]
+        public void Add_Tokens()
+        {
+            var user = new User(GlobalSettings, "Test", "test@test.com", "test", "helloworldtest");
```
Comment (bergmania) [inline src/Umbraco.Tests.Integration/Umbraco.Infrastructure/Services/ExternalLoginServiceTests.cs:209]: I know it is just copy/pasted, but when we don't use any of this dummy parameters, we really should use our builder. This suggestion can be done for the entire test class

```suggestion
            var user = new UserBuilder().Build();
```

**Review by bergmania — APPROVED** [review]
Looks good.. Only found small things I will fix up for you.

**Comment by bergmania** [comment]
@Shazwazza , if you are happy with my changes, feel free to merge.. My tests did not find anything unexpected.

**Comment by Shazwazza** [comment]
@bergmania awesome 💪

**Pull request merged**